### PR TITLE
Connection pool lease fix

### PR DESCRIPTION
**Pull request by neunhoef**
This fixes a bug in the connectionPool implementation, which could
lead to the fact that basically no connections have been reused.

Under high load and high parallelism it could happen that a connection
was leased from fuerte but could then not be leased in the connection
pool. In this case the lease from fuerte was not returned and thus the
connection could no longer be used. In k8s and with TLS we have observed
situations where the timing was so that this built up and created
thousands of connections in a very short period of time.

This let performance suffer seriously.

The solution is essentially to release the lease immediately in the
above mentioned case.

We also added some more metrics to make this behaviour better
observable.

### Scope & Purpose

*(Please describe the changes in this PR for reviewers)*

- [*] :hankey: Bugfix 
- [ ] :pizza: New feature 
- [*] :hammer: Refactoring 
- [*] :book: CHANGELOG entry made
- [*] :muscle: The behavior in this PR was *manually tested*
- [ ] :computer: The behavior change can be verified via automatic tests

#### Backports:

- [ ] No backports required
- [*] Backports required for: 3.6

### Testing & Verification

*(Please pick either of the following options)*

- [*] This change is a trivial rework / code cleanup without any test coverage.

### CODE REVIEW

File: arangod/Network/ConnectionPool.cpp [hunk 1]
```diff
@@ -37,8 +38,24 @@ namespace network {
 using namespace arangodb::fuerte::v1;
 
 ConnectionPool::ConnectionPool(ConnectionPool::Config const& config)
-    : _config(config), 
-      _loop(config.numIOThreads, config.name) {
+    : _config(config),
+      _loop(config.numIOThreads, config.name),
+      _totalConnectionsInPool(
+        _config.clusterInfo->server().getFeature<arangodb::MetricsFeature>().gauge(
+          std::string("arangodb_connection_pool_nr_conns_") + _config.name, uint64_t(0), "Total number of connection in pool")),
```
Comment (jsteemann) [inline arangod/Network/ConnectionPool.cpp:45]:
```suggestion
          std::string("arangodb_connection_pool_nr_conns_") + _config.name, uint64_t(0), "Current number of connections in pool")),
```

File: arangod/Network/ConnectionPool.cpp [hunk 2]
```diff
@@ -37,8 +38,24 @@ namespace network {
...
+      _successSelect(
+        _config.clusterInfo->server().getFeature<arangodb::MetricsFeature>().counter(
+          std::string("arangodb_connection_pool_success_select_") + _config.name, 0, "Success select lease")),
```
Comment (jsteemann) [inline arangod/Network/ConnectionPool.cpp:48]:
```suggestion
          std::string("arangodb_connection_pool_success_select_") + _config.name, 0, "Total number of successful connection leases")),
```

**Review by kvahed — APPROVED** [review]
LGTM

File: arangod/Network/ConnectionPool.cpp [hunk 3]
```diff
@@ -37,8 +38,24 @@ namespace network {
...
+      _noSuccessSelect(
+        _config.clusterInfo->server().getFeature<arangodb::MetricsFeature>().counter(
+          std::string("arangodb_connection_pool_no_success_select_") + _config.name, 0, "No success select lease")),
```
Comment (jsteemann) [inline arangod/Network/ConnectionPool.cpp:51]:
```suggestion
          std::string("arangodb_connection_pool_no_success_select_") + _config.name, 0, "Total number of failed connection leases")),
```

File: arangod/Network/ConnectionPool.cpp [hunk 4]
```diff
@@ -231,15 +253,22 @@ ConnectionPtr ConnectionPool::selectConnection(std::string const& endpoint,
         // next check against the number of requests in flight
         if (c->fuerte->requestsLeft() <= limit) {
           c->lastLeased = std::chrono::steady_clock::now();
+          _successSelect++;
+          _leaseHistMSec.count(
+            duration<float,std::milli>(high_resolution_clock::now()-start).count());
```
Comment (jsteemann) [inline arangod/Network/ConnectionPool.cpp]:
```suggestion
            duration<float, std::milli>(high_resolution_clock::now() - start).count());
```

File: arangod/Network/ConnectionPool.cpp [hunk 5]
```diff
@@ -231,15 +253,22 @@ ConnectionPtr ConnectionPool::selectConnection(std::string const& endpoint,
         // next check against the number of requests in flight
         if (c->fuerte->requestsLeft() <= limit) {
           c->lastLeased = std::chrono::steady_clock::now();
+          _successSelect++;
+          _leaseHistMSec.count(
+            duration<float,std::micro>(high_resolution_clock::now()-start).count());
```
Comment (jsteemann) [inline arangod/Network/ConnectionPool.cpp:258]:
```suggestion
            duration<float, std::micro>(high_resolution_clock::now() - start).count());
```

File: arangod/Network/ConnectionPool.cpp [hunk 6]
```diff
@@ -250,6 +279,10 @@ ConnectionPtr ConnectionPool::selectConnection(std::string const& endpoint,
   std::shared_ptr<fuerte::Connection> fuerte = createConnection(builder);
   auto c = std::make_shared<Context>(fuerte, std::chrono::steady_clock::now(), 1 /* leases*/);
   bucket.list.push_back(c);
+  _totalConnectionsInPool += 1;
+  
+  _leaseHistMSec.count(
+    duration<float,std::milli>(high_resolution_clock::now()-start).count());
```
Comment (jsteemann) [inline arangod/Network/ConnectionPool.cpp:285]:
```suggestion
    duration<float, std::milli>(high_resolution_clock::now() - start).count());
```

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/13123/

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/13124/

File: arangod/Network/ConnectionPool.cpp (hunk 4 above)
Comment (neunhoef) [reply arangod/Network/ConnectionPool.cpp]: Done

File: arangod/Network/ConnectionPool.cpp [hunk 7]
```diff
@@ -199,8 +217,12 @@ std::shared_ptr<fuerte::Connection> ConnectionPool::createConnection(fuerte::Con
 
 ConnectionPtr ConnectionPool::selectConnection(std::string const& endpoint,
                                                ConnectionPool::Bucket& bucket) {
+
+  using namespace std::chrono;
   std::lock_guard<std::mutex> guard(bucket.mutex);
 
+  auto start = high_resolution_clock::now();
```
Comment (jsteemann) [inline arangod/Network/ConnectionPool.cpp:224]: from https://en.cppreference.com/w/cpp/chrono/high_resolution_clock:

> The high_resolution_clock is not implemented consistently across different standard library implementations, and its use should be avoided. It is often just an alias for std::chrono::steady_clock or std::chrono::system_clock, but which one it is depends on the library or configuration. When it is a system_clock, it is not monotonic (e.g., the time can go backwards). For example, for gcc's libstdc++ it is system_clock, for MSVC it is steady_clock, and for clang's libc++ it depends on configuration.

Comment (jsteemann) [reply arangod/Network/ConnectionPool.cpp:224]: Let's rather use `steady_clock` here, so the clock cannot go backwards and negative durations are avoided.

File: arangod/Network/ConnectionPool.cpp [hunk 8]
```diff
@@ -231,15 +253,22 @@ ConnectionPtr ConnectionPool::selectConnection(std::string const& endpoint,
         // next check against the number of requests in flight
         if (c->fuerte->requestsLeft() <= limit) {
           c->lastLeased = std::chrono::steady_clock::now();
+          _successSelect++;
+          _leaseHistMSec.count(
+            duration<float, std::micro>(high_resolution_clock::now() - start).count());
```
Comment (jsteemann) [inline arangod/Network/ConnectionPool.cpp:258]: If we moved the durations to a `steady_clock`, we could avoid one call to `now()` here, and just reuse the result.

Comment (neunhoef) [reply arangod/Network/ConnectionPool.cpp:258]: Indeed. Will do.

File: arangod/Network/ConnectionPool.cpp (hunk 7 above)
Comment (neunhoef) [reply arangod/Network/ConnectionPool.cpp:224]: Will do.

**Review by jsteemann — APPROVED** [review]

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/13150/

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/13198/

**Comment by neunhoef** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/13206/

**Comment by KVS85** [comment]
Tests blue except unrelated failures.

**Pull request merged**